lta/bundler.py

```diff
@@ -0,0 +1,324 @@
+# bundler.py
+"""Module to implement the Bundler component of the Long Term Archive."""
+
+import asyncio
+from datetime import datetime
+import hashlib
+import json
+from logging import Logger
+import logging
+import os
+import platform
+import shutil
+import sys
+from typing import Any, Dict, List
+from uuid import uuid4
+from zipfile import ZIP_STORED, ZipFile
+
+from rest_tools.client import RestClient  # type: ignore
+from urllib.parse import urljoin
+
+from .config import from_environment
+from .log_format import StructuredFormatter
+
+EXPECTED_CONFIG = {
+    "BUNDLER_NAME": f"{platform.node()}-bundler",
+    "BUNDLER_SITE_SOURCE": None,
+    "HEARTBEAT_PATCH_RETRIES": "3",
+    "HEARTBEAT_PATCH_TIMEOUT_SECONDS": "30",
+    "HEARTBEAT_SLEEP_DURATION_SECONDS": "60",
+    "LTA_REST_TOKEN": None,
+    "LTA_REST_URL": None,
+    "LTA_SITE_CONFIG": "etc/site.json",
+    "OUTBOX_PATH": None,
+    "WORK_RETRIES": "3",
+    "WORK_SLEEP_DURATION_SECONDS": "300",
+    "WORK_TIMEOUT_SECONDS": "30",
+    "WORKBOX_PATH": None,
+}
+
+HEARTBEAT_STATE = [
+    "last_work_begin_timestamp",
+    "last_work_end_timestamp",
+    "lta_ok"
+]
+
+def now() -> str:
+    """Return string timestamp for current time, to the second."""
+    return datetime.utcnow().isoformat(timespec='seconds')
+
+# Adapted from: https://stackoverflow.com/a/44873382
+def sha512sum(filename: str) -> str:
+    """Compute the SHA512 hash of the supplied filename."""
+    h = hashlib.sha512()
+    b = bytearray(128*1024)
+    mv = memoryview(b)
+    with open(filename, 'rb', buffering=0) as f:
+        # Known issue with MyPy: https://github.com/python/typeshed/issues/2166
+        for n in iter(lambda: f.readinto(mv), 0):  # type: ignore
+            h.update(mv[:n])
+    return h.hexdigest()
```

Comment:
`sha512sum` should probably be factored out into a util file, since we'll need it in a few other places.

Reply:
Refactored out into `lta/crypto.py` and added a unit test for the function.